### PR TITLE
General solution to supporing flexbox, grid and any other layout with "gaps"

### LINKED ISSUE #244: Find a more general solution to calculate positions during drag and drop on elements with special spacing
Right now, during a drag and drop or elements in the page, when doing the math to calculate the position of the placeholder or the rearranged elements, we're taking into account the x/y coordinates, the height/width and the marginLeft/marginTop.

However, that is only true for very basic layouts. As soon as we use row-gaps or flexbox with space-between, those calculations are not correct, because there's space unaccounted for.

We need to find a more general solution that works for those cases.

### PR DESCRIPTION
**Pull request by cibernox**
Fixes #244 (maybe)

### CODE REVIEW

**Comment by cibernox** [comment]
I've created this page with all kinds of examples to try to easily test all edge cases:
```
<!-- 1. Row using Margins -->
<div class="mb-8">
  <h2 class="text-lg font-semibold mb-4">1. Row using Margins</h2>
  <div class="flex">
    <div class="mr-4 p-2 border rounded">Item 1<br />Small text.</div>
    <div class="mr-4 p-2 border rounded">Item 2<br />Small text.</div>
    <div class="mr-4 p-2 border rounded">Item 4<br />Small text.</div>
    <div class="mr-4 p-2 border rounded flex-grow">Item 3<br />Wider element text.</div>
    <div class="mr-4 p-2 border rounded">Item 5<br />Small text.</div>
  </div>
</div>
<!-- 2. Vertical using Margins -->
<div class="mb-8">
  <h2 class="text-lg font-semibold mb-4">2. Vertical using Margins</h2>
  <div>
    <div class="mb-4 p-2 border rounded">Item 1<br />Small text.</div>
    <div class="mb-4 p-2 border rounded">Item 2<br />Small text.</div>
    <div class="mb-4 p-2 border rounded">Item 4<br />Small text.</div>
    <div class="mb-4 p-2 border rounded">Item 3<br />Small text. <br /> But taller</div>
    <div class="mb-4 p-2 border rounded">Item 5<br />Small text.</div>
  </div>
</div>
<!-- 3. Row using Flexbox -->
<div class="mb-8">
  <h2 class="text-lg font-semibold mb-4">3. Row using Flexbox</h2>
  <div class="flex space-x-4">
    <div class="p-2 border rounded">Item 1<br />Small text.</div>
    <div class="p-2 border rounded">Item 2<br />Small text.</div>
    <div class="p-2 border rounded">Item 4<br />Small text.</div>
    <div class="p-2 border rounded flex-grow">Item 3<br />Wider element text.</div>
    <div class="p-2 border rounded">Item 5<br />Small text.</div>
  </div>
</div>
<!-- 4. Vertical using Flexbox -->
<div class="mb-8">
  <h2 class="text-lg font-semibold mb-4">4. Vertical using Flexbox</h2>
  <div class="flex flex-col space-y-4">
    <div class="p-2 border rounded">Item 1<br />Small text.</div>
    <div class="p-2 border rounded">Item 2<br />Small text.</div>
    <div class="p-2 border rounded">Item 3<br />Small text. <br /> But taller</div>
    <div class="p-2 border rounded">Item 4<br />Small text.</div>
    <div class="p-2 border rounded">Item 5<br />Small text.</div>
  </div>
</div>
<!-- 5. Row using CSS Grid -->
<div class="mb-8">
  <h2 class="text-lg font-semibold mb-4">5. Row using CSS Grid (Single Row)</h2>
  <div class="grid grid-cols-6 gap-4">
    <div class="p-2 border rounded">Item 1<br />Small text.</div>
    <div class="p-2 border rounded">Item 2<br />Small text.</div>
    <div class="p-2 border rounded col-span-2">Item 3<br />Wider element text.</div>
    <div class="p-2 border rounded">Item 4<br />Small text.</div>
    <div class="p-2 border rounded">Item 5<br />Small text.</div>
  </div>
</div>
<!-- 6. Vertical using CSS Grid -->
<div class="mb-8">
  <h2 class="text-lg font-semibold mb-4">6. Vertical using CSS Grid</h2>
  <div class="grid gap-4 grid-rows-6">
    <div class="p-2 border rounded">Item 1<br />Small text.</div>
    <div class="p-2 border rounded">Item 2<br />Small text.</div>
    <div class="p-2 border rounded row-span-2">Item 3<br />Small text. <br /> But taller</div>
    <div class="p-2 border rounded">Item 4<br />Small text.</div>
    <div class="p-2 border rounded">Item 5<br />Small text.</div>
  </div>
</div>
<!-- 7. 2x3 Grid of Elements -->
<div class="mb-8">
  <h2 class="text-lg font-semibold mb-4">7. 2x3 Grid of Elements</h2>
  <div class="grid grid-cols-2 gap-4">
    <div class="p-2 border rounded">Item 1<br />Small text.</div>
    <div class="p-2 border rounded">Item 2<br />Small text.<br /> But taller</div>
    <div class="p-2 border rounded">Item 3<br />Small text.</div>
    <div class="p-2 border rounded">Item 4<br />Small text.</div>
    <div class="p-2 border rounded">Item 5<br />Small text.</div>
    <div class="p-2 border rounded">Item 6<br />Small text.</div>
  </div>
</div>

```

**Comment by cibernox** [comment]
https://github.com/user-attachments/assets/8629129c-fb2d-438d-8a08-0cb0de4a84bf

**Comment by cibernox** [comment]
This looks very promising:

#### Before:
- ✅ Row using Margins
- ✅ Vertical using Margins
- ❌ Row using Flexbox
- ❌Vertical using Flexbox
- ✅ Row using CSS Grid (Single Row
- ✅ Vertical using CSS Grid
- ❌ 2x3 Grid of Elements

#### After:
- ✅ Row using Margins
- ✅ Vertical using Margins
- ✅ Row using Flexbox
- ✅ Vertical using Flexbox
- ✅ Row using CSS Grid (Single Row
- ✅ Vertical using CSS Grid
- ❌ 2x3 Grid of Elements

I think that drag and drop when elements flow both horizontally and vertically should be tackled in another PR as it's quite complex.

**Review by cbroeren — APPROVED** [review]
I wasn't able to test it myself locally yet, but it looks really smooth from your screen recording 💯

File: assets/svelte/components/SelectedElementFloatingMenu/DragMenuOption.svelte [hunk 1]
```diff
@@ -243,16 +243,34 @@
     let i = 0
     if (dragDirection === "vertical") {
       while (i < newInfos.length && i < newIndex) {
-        distance += newInfos[i].height + newInfos[i].marginTop + newInfos[i].marginBottom
+        let gap = 0
```
Comment (cibernox) [inline assets/svelte/components/SelectedElementFloatingMenu/DragMenuOption.svelte:246]: Instead of assuming the "void" between elements comes determined by the margin (which is plainly wrong in anything but the most basic layout techniques), we measure the distance from the rightmost edge of one elements and the leftmost edge of the next.

That is a generic solution that works for marging, flebox, css-grid and probably any other layout approach.

**Comment by leandrocp** [comment]
> I've created this page with all kinds of examples to try to easily test all edge cases:

This is super useful. I'm gonna add it to the dev app in beacon.

**Review by leandrocp — APPROVED** [review]
Working pretty well from the user perspective. I'll leave the code review for @cbroeren

**Comment by cibernox** [comment]
@leandrocp maybe you could add the JS formatting as a pre-commit git hook? So you don't have to manually do it ever again.

**Comment by leandrocp** [comment]
> @leandrocp maybe you could add the JS formatting as a pre-commit git hook? So you don't have to manually do it ever again.

I like the idea of auto formatting but I'm not a big fan of git hooks (too problematic). So we can use a CI action instead, for example https://github.com/BeaconCMS/beacon_live_admin/pull/247/commits/340340ca0cc6de9096481e8a85835b8364e2f797

https://github.com/BeaconCMS/beacon_live_admin/pull/247

**Comment by cibernox** [comment]
As long as it's automated, it works for me.
